# <center>Assignment 1
    Submitted by: DIBYA RANJAN DASH
    BRANCH: CEN
    SIC: 20BECB77

***

# Part A: Data Preprocessing

##  1. Import Libraries

In [90]:
import pandas as pd
import numpy as np
import matplotlib.pylab as plt

## 2. Read the data file

In [91]:
dataset = pd.read_csv('vehicle.data')

In [92]:
dataset

vhigh vhigh.1      2   2.1  small   low  unacc
0     vhigh   vhigh      2     2  small   med  unacc
1     vhigh   vhigh      2     2  small  high  unacc
2     vhigh   vhigh      2     2    med   low  unacc
3     vhigh   vhigh      2     2    med   med  unacc
4     vhigh   vhigh      2     2    med  high  unacc
...     ...     ...    ...   ...    ...   ...    ...
1722    low     low  5more  more    med   med   good
1723    low     low  5more  more    med  high  vgood
1724    low     low  5more  more    big   low  unacc
1725    low     low  5more  more    big   med   good
1726    low     low  5more  more    big  high  vgood

[1727 rows x 7 columns]

## 3. To check if there is some missing data

In [93]:
dataset.isnull().sum()

vhigh      0
vhigh.1    0
2          0
2.1        0
small      0
low        0
unacc      0
dtype: int64

## 4. To create feature matrix(X) and dependent variable vector(Y)

In [94]:
X = dataset.iloc[:,:-1].values
Y = dataset.iloc[:,-1].values

In [95]:
X

array([['vhigh', 'vhigh', '2', '2', 'small', 'med'],
       ['vhigh', 'vhigh', '2', '2', 'small', 'high'],
       ['vhigh', 'vhigh', '2', '2', 'med', 'low'],
       ...,
       ['low', 'low', '5more', 'more', 'big', 'low'],
       ['low', 'low', '5more', 'more', 'big', 'med'],
       ['low', 'low', '5more', 'more', 'big', 'high']], dtype=object)

In [96]:
Y

array(['unacc', 'unacc', 'unacc', ..., 'unacc', 'good', 'vgood'],
      dtype=object)

## 5. Encoding

### Encoding Feature matrix using Ordinal Encoder

In [97]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OrdinalEncoder
enc = OrdinalEncoder()
X = np.array(enc.fit_transform(X))


In [98]:
X

array([[3., 3., 0., 0., 2., 2.],
       [3., 3., 0., 0., 2., 0.],
       [3., 3., 0., 0., 1., 1.],
       ...,
       [1., 1., 3., 2., 0., 1.],
       [1., 1., 3., 2., 0., 2.],
       [1., 1., 3., 2., 0., 0.]])

### Encoding Dependent variable using Label Encoder

In [99]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
Y = np.array(le.fit_transform(Y))

In [100]:
Y

array([2, 2, 2, ..., 2, 1, 3])

## 6. Splitting to dataset into training dataset and testing dataset

In [101]:
from sklearn.model_selection import train_test_split
Xtrain,Xtest,Ytrain,Ytest = train_test_split(X,Y,test_size = 0.2,random_state=1)

In [102]:
Xtrain

array([[1., 2., 2., 1., 1., 2.],
       [0., 0., 3., 1., 1., 0.],
       [3., 2., 3., 0., 1., 1.],
       ...,
       [2., 2., 0., 1., 0., 0.],
       [3., 2., 0., 2., 2., 0.],
       [2., 0., 3., 1., 2., 1.]])

In [103]:
Xtest

array([[1., 0., 0., 2., 1., 1.],
       [0., 0., 1., 2., 0., 0.],
       [0., 0., 3., 0., 1., 0.],
       ...,
       [3., 1., 2., 1., 2., 1.],
       [3., 3., 1., 2., 1., 0.],
       [3., 3., 3., 1., 1., 0.]])

# a) : Bulding Logistic Regression Model

## 1. Training Logistic Regression Model

In [104]:
from sklearn.linear_model import LogisticRegression
LoR = LogisticRegression(random_state=1)
LoR.fit(Xtrain,Ytrain)

LogisticRegression(random_state=1)

## 2. Testing Logistic Regression Model

In [105]:
Yestimated = LoR.predict(Xtest)
print(np.concatenate((Yestimated.reshape(len(Yestimated),1),Ytest.reshape(len(Ytest),1)),1))


[[2 2]
 [3 0]
 [2 2]
 [2 2]
 [2 2]
 [2 0]
 [0 3]
 [2 2]
 [2 0]
 [2 3]
 [0 0]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 0]
 [2 0]
 [0 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 0]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 0]
 [2 3]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [0 3]
 [2 0]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [0 2]
 [2 2]
 [0 0]
 [0 2]
 [2 1]
 [2 0]
 [2 1]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [0 0]
 [2 0]
 [2 2]
 [2 2]
 [2 2]
 [3 3]
 [2 2]
 [2 1]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 0]
 [2 2]
 [2 0]
 [2 2]
 [0 1]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [0 0]
 [2 0]
 [0 3]
 [2 2]
 [2 2]
 [2 0]
 [0 3]
 [0 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [0 0]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [0 2]
 [2 2]
 [2 2]
 [2 2]
 [2 0]
 [2 2]
 [2 2]
 [2 2]
 [2 0]
 [2 2]
 [2 1]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [0 2]
 [2 2]
 [2 3]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [0 2]
 [2 2]
 [0 3]
 [2 2]
 [2 2]
 [2 0]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 0]

## 3. Confusion Matrix

In [106]:
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(Ytest,Yestimated)
print('Confusion Matrix is: \n\n',cm)


Confusion Matrix is: 

 [[ 15   0  57   1]
 [  2   0  11   0]
 [ 22   0 217   0]
 [  9   0  10   2]]


## 4. Accuracy Score

In [107]:
from sklearn.metrics import accuracy_score
print('The Accuracy Score of the Model is: ',accuracy_score(Ytest,Yestimated))


The Accuracy Score of the Model is:  0.6763005780346821


## 5. Classification Report

In [108]:
from sklearn.metrics import classification_report
print(classification_report(Ytest , Yestimated))

              precision    recall  f1-score   support

           0       0.31      0.21      0.25        73
           1       0.00      0.00      0.00        13
           2       0.74      0.91      0.81       239
           3       0.67      0.10      0.17        21

    accuracy                           0.68       346
   macro avg       0.43      0.30      0.31       346
weighted avg       0.61      0.68      0.62       346



C:\Users\dibya\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\dibya\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\dibya\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


# b): Building Naive Bye's Model

## 1. Training Naive Bye's Model

In [109]:
from sklearn.naive_bayes import GaussianNB
NB = GaussianNB()
NB.fit(Xtrain,Ytrain)

GaussianNB()

## 2. Testing Naive Bye's Model

In [110]:
Yestimated = NB.predict(Xtest)
print(np.concatenate((Yestimated.reshape(len(Yestimated),1),Ytest.reshape(len(Ytest),1)),1))

[[2 2]
 [3 0]
 [3 2]
 [2 2]
 [2 2]
 [3 0]
 [3 3]
 [2 2]
 [2 0]
 [3 3]
 [3 0]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [3 0]
 [2 0]
 [3 2]
 [2 2]
 [2 2]
 [3 2]
 [2 2]
 [2 0]
 [3 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 0]
 [3 3]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [3 3]
 [3 0]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [3 2]
 [3 0]
 [3 2]
 [2 1]
 [2 0]
 [0 1]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [3 2]
 [0 0]
 [2 0]
 [2 2]
 [2 2]
 [2 2]
 [3 3]
 [2 2]
 [3 1]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 0]
 [2 2]
 [2 0]
 [2 2]
 [0 1]
 [2 2]
 [2 2]
 [2 2]
 [3 2]
 [3 0]
 [3 0]
 [3 3]
 [2 2]
 [2 2]
 [2 0]
 [3 3]
 [2 2]
 [3 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [0 2]
 [0 0]
 [3 2]
 [2 2]
 [2 2]
 [2 2]
 [3 2]
 [2 2]
 [2 2]
 [2 2]
 [3 0]
 [2 2]
 [2 2]
 [2 2]
 [3 0]
 [2 2]
 [0 1]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [3 3]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [3 2]
 [2 2]
 [2 2]
 [2 2]
 [3 3]
 [2 2]
 [2 2]
 [2 0]
 [3 2]
 [2 2]
 [2 2]
 [2 2]
 [3 0]

## 3. Confusion Matrix

In [111]:
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(Ytest,Yestimated)
print('Confusion Matrix is: \n\n',cm)

Confusion Matrix is: 

 [[  7   0  31  35]
 [  3   0   5   5]
 [  5   0 203  31]
 [  0   0   0  21]]


## 4. Accuracy Score

In [112]:
from sklearn.metrics import accuracy_score
print('The Accuracy Score of the Model is: ',accuracy_score(Ytest,Yestimated))

The Accuracy Score of the Model is:  0.6676300578034682


## 5. Classification Report

In [113]:
from sklearn.metrics import classification_report
print(classification_report(Ytest , Yestimated))

              precision    recall  f1-score   support

           0       0.47      0.10      0.16        73
           1       0.00      0.00      0.00        13
           2       0.85      0.85      0.85       239
           3       0.23      1.00      0.37        21

    accuracy                           0.67       346
   macro avg       0.39      0.49      0.35       346
weighted avg       0.70      0.67      0.64       346



C:\Users\dibya\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\dibya\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\dibya\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


# c) Decision Tree Classifier

## 1. Training Decision Tree Classifier Model

In [114]:
from sklearn.tree import DecisionTreeClassifier

In [115]:
DT = DecisionTreeClassifier(criterion='entropy', random_state = 0)
DT.fit(Xtrain,Ytrain)

DecisionTreeClassifier(criterion='entropy', random_state=0)

## 2. Testing Decision Tree Classifier Model

In [116]:
Yestimated = DT.predict(Xtest)

In [117]:
print(np.concatenate((Yestimated.reshape(len(Yestimated),1),Ytest.reshape(len(Ytest),1)),1))

[[2 2]
 [0 0]
 [2 2]
 [2 2]
 [2 2]
 [0 0]
 [3 3]
 [2 2]
 [0 0]
 [3 3]
 [0 0]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [0 0]
 [0 0]
 [0 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [0 0]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [0 0]
 [3 3]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [3 3]
 [0 0]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [0 0]
 [2 2]
 [1 1]
 [2 0]
 [1 1]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [0 0]
 [2 0]
 [0 2]
 [2 2]
 [2 2]
 [3 3]
 [2 2]
 [1 1]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [0 0]
 [2 2]
 [0 0]
 [2 2]
 [1 1]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [0 0]
 [0 0]
 [3 3]
 [2 2]
 [2 2]
 [0 0]
 [3 3]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [0 0]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [0 0]
 [2 2]
 [2 2]
 [2 2]
 [0 0]
 [2 2]
 [1 1]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [3 3]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [3 3]
 [2 2]
 [2 2]
 [0 0]
 [2 2]
 [2 2]
 [2 2]
 [2 2]
 [0 0]

## 3. Confusion Matrix

In [118]:
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(Ytest,Yestimated)
print('Confusion Matrix is: \n\n',cm)

Confusion Matrix is: 

 [[ 71   0   2   0]
 [  2  11   0   0]
 [  3   0 236   0]
 [  1   0   0  20]]


## 4. Accuracy Score

In [119]:
from sklearn.metrics import accuracy_score
print('The Accuracy Score of the Model is: ',accuracy_score(Ytest,Yestimated))

The Accuracy Score of the Model is:  0.976878612716763


## 5. Classification Report

In [120]:
from sklearn.metrics import classification_report
print(classification_report(Ytest , Yestimated))

              precision    recall  f1-score   support

           0       0.92      0.97      0.95        73
           1       1.00      0.85      0.92        13
           2       0.99      0.99      0.99       239
           3       1.00      0.95      0.98        21

    accuracy                           0.98       346
   macro avg       0.98      0.94      0.96       346
weighted avg       0.98      0.98      0.98       346



# d) Random Forest Classifier

## 1. Training Random Forest Classifier

In [121]:
 from sklearn.ensemble import RandomForestClassifier

In [122]:
rfc = RandomForestClassifier(n_estimators=40,criterion='entropy',random_state=0)

In [123]:
rfc.fit(Xtrain,Ytrain)

RandomForestClassifier(criterion='entropy', n_estimators=40, random_state=0)

## 2. Testing Random Forest Classifier

In [124]:
Yestimated = rfc.predict(Xtest)

In [125]:
Yestimated

array([2, 0, 2, 2, 2, 0, 3, 2, 0, 3, 0, 2, 2, 2, 2, 2, 0, 0, 0, 2, 2, 2,
       2, 0, 2, 2, 2, 2, 2, 0, 3, 2, 2, 2, 2, 2, 2, 2, 3, 0, 2, 2, 2, 2,
       2, 2, 0, 0, 1, 2, 1, 2, 2, 2, 2, 2, 0, 2, 0, 2, 2, 3, 2, 1, 2, 2,
       2, 2, 2, 2, 0, 2, 0, 2, 1, 2, 2, 2, 2, 0, 0, 3, 2, 2, 0, 3, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 0, 2, 2, 2, 0, 2, 2, 2, 2, 2, 2, 2,
       2, 0, 2, 2, 2, 0, 2, 1, 2, 2, 2, 2, 2, 2, 3, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 3, 2, 2, 0, 2, 2, 2, 2, 0, 0, 2, 2, 2, 0, 2, 2, 2, 0, 0, 2,
       2, 0, 0, 2, 2, 2, 2, 2, 0, 2, 2, 2, 2, 2, 1, 2, 0, 2, 2, 0, 0, 0,
       2, 2, 0, 2, 2, 2, 2, 2, 0, 3, 2, 2, 2, 2, 0, 2, 2, 3, 2, 2, 0, 2,
       3, 2, 0, 2, 0, 0, 2, 2, 2, 3, 2, 0, 0, 2, 2, 1, 2, 3, 2, 2, 2, 3,
       2, 2, 0, 0, 2, 2, 2, 2, 2, 0, 2, 2, 2, 0, 2, 0, 2, 2, 2, 2, 2, 2,
       0, 0, 2, 0, 2, 2, 2, 0, 2, 3, 1, 2, 0, 0, 0, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 0, 2, 2, 2, 0, 2, 2, 2, 2, 0, 2, 0, 1, 2, 2, 2, 1, 2, 0,
       0, 2, 2, 2, 2, 2, 2, 2, 0, 2, 0, 3, 2, 2, 2,

## 3. Confusion Matrix

In [126]:
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(Ytest,Yestimated)
print('Confusion Matrix is: \n\n',cm)

Confusion Matrix is: 

 [[ 68   0   5   0]
 [  2  11   0   0]
 [  5   0 234   0]
 [  1   0   0  20]]


## 4. Accuracy Score

In [127]:
from sklearn.metrics import accuracy_score
accuracy_score(Ytest,Yestimated)

0.9624277456647399

## 5. Classification Report

In [128]:
from sklearn.metrics import classification_report
print(classification_report(Ytest , Yestimated))

              precision    recall  f1-score   support

           0       0.89      0.93      0.91        73
           1       1.00      0.85      0.92        13
           2       0.98      0.98      0.98       239
           3       1.00      0.95      0.98        21

    accuracy                           0.96       346
   macro avg       0.97      0.93      0.95       346
weighted avg       0.96      0.96      0.96       346



***

### So the Accuracy of all the models are as folllows: 

 >a) Logistic Regression Model: 0.6763

 >b) Naive Bye's Model: 0.6676

 >c) Decision Tree Classifier Model: 0.9768

 >d) Random Forest Classifier Model: 0.9624

 Here the __Decision Tree Algorithm__ and __Random Forest Classifier__ has Accuracy of __97%__ and __96%__ respectively for __Vehicle.data__